File: cpp/affine_point.cpp
#include "field.cpp"
#include <iostream>


template<mpz_t a, mpz_t b, mpz_t mod>
class AffinePoint
{
private:
    typedef Field<mod> Number;
    Number x;
    Number y;
public:
    AffinePoint(mpz_t x, mpz_t y){
       this->x = Number(x);
       this->y = Number(y);
    }
    AffinePoint(mpz_class x, mpz_class y){
       this->x = Number(x);
       this->y = Number(y);
    }
    AffinePoint(long long x, long long y){
       this->x = Number(x);
       this->y = Number(y);
    }
    AffinePoint(int x, int y){
       this->x = Number(x);
       this->y = Number(y);
    }
    AffinePoint(Number x, Number y){
       this->x = x;
       this->y = y;
    }

    bool is_zero(){
        return x == Number(0);
    }

    Number getX(){
        return x;
    }

    Number getY(){
        return y;
    }


    AffinePoint dbl() {
        if (this->is_zero() || this->y == Number(0)){
            return AffinePoint(0,0);
        }
        else{ 
            Number temp = (x * x  * Number(3) + Number(a)) / (y * Number(2));
            Number rx = temp * temp - x * Number(2);
            Number ry = temp * (x - rx) - y;
            return AffinePoint(rx, ry);
        }
    }

    AffinePoint operator+ (AffinePoint const &other) {
        if(this->is_zero()) {
            return other;
        }
        else if (Number(other.x) == Number(0)) {
            return *this;
        }
        else if (this->x == other.x) {
            if (this->y == other.y){
                return this->dbl();
            }
            else{
                return AffinePoint(0,0);
            }
        }
        else {
            Number temp = (Number(other.y) - y) / (Number(other.x) - x);
            Number rx = temp * temp - x - other.x;
            Number ry = temp * (x - rx) - y;
            return AffinePoint(rx, ry);
        }
    }

    AffinePoint operator* (mpz_class n){
        AffinePoint result = AffinePoint(0,0);
        AffinePoint temp = *this;
        while(n != 0){
            if (n % 2 != 0){
                result = result + temp;
            }
            temp =  temp.dbl();
            n = n >> 1;
        }
        return result;
    }

    bool operator== (AffinePoint const &other){
         if(this->is_zero() || Number(other.x) == Number(0)){
             return this->is_zero() && Number(other.x) == Number(0);
         }
         else {
             return (x == other.x) && (y == other.y);
         }
    }
    
    void print() {
        std::cout << "(" << x.getValue()<< ", " << y.getValue() <<")"<<std::endl; // actual output done here
 
    }
     
};

File: affine_point.py
from finite_field import F



def AffinePoint(a, b, mod):
    class AffinePoint:
        __slots__ = ['field', 'x', 'y']
        def __init__(self, x, y):
            self.field = F(mod)
            if isinstance(x, int):
                self.x = self.field(x)
            else:
                self.x = x
            if isinstance(y, int):
                self.y = self.field(y)
            else:
                self.y = y            

        def is_zero(self):
            return self.x is None

        def __add__(self, other):
            if self.is_zero():
                return other
            elif other.is_zero():
                return self
            elif self.x == other.x:
                if self.y == other.y:
                    return self.double()
                else:
                    return AffinePoint(None, None)
            else:
                temp = (other.y - self.y) / (other.x - self.x)
                rx = temp * temp - self.x - other.x
                ry = - self.y + temp * (self.x - rx) 
                return AffinePoint(rx, ry)
            
        def double(self):
            if self.is_zero() or self.y == self.field(0):
                return AffinePoint(None, None)
            else:
                temp = (self.x * self.x * self.field(3) + self.field(a)) / (self.y * self.field(2))
                rx = temp * temp - self.x * self.field(2)
                ry = temp * (self.x - rx) - self.y
                return AffinePoint(rx, ry)

        def __neg__(self):
            if self.is_zero():
                return self
            else:
                return AffinePoint(self.x, -self.y)

        def __mul__(self, n):
            result = AffinePoint(None, None)
            temp = self
            while n != 0:
                if n & 1 != 0:
                    result += temp
                temp = temp.double()
                n >>= 1
            return result

        def __str__(self):
            if self.is_zero():
                return "(Zero)"
            else:
                return f"({self.x}, {self.y})"

        __rmul__ = __mul__
        __repr__ = __str__

        def __eq__(self, other):
            if self.is_zero() or other.is_zero():
                return self.is_zero() and other.is_zero()
            else:
                return (self.x, self.y) == (other.x, other.y)

    return AffinePoint
